### DISCUSSION TITLE
Future Plans

### DISCUSSION THREAD

**Discussion by twlite** [Androz2091/discord-player#989] [General]
Hello everyone. Sorry if I have a bad English (I'm not a native speaker) 😛. I am the maintainer of this project, @Androz2091 being busy can't handle this project along with his other projects at the same time. Due to this, I am trying to update this library whenever possible to reduce his workload. Note that I am also not free and I have many work to do myself. I mostly update this library when I am free/have time. With that being said, I want to list out some future plans for this library.

Discord Player is currently on version 5. Everything started from **[Jan 11 2020](https://github.com/Androz2091/discord-player/commit/d7eae5b1697ea3dfb0e81a0d9f806c3c63a90e4d)** and the community has been growing ever since. According to GitHub reports, over 3.2K+ projects depend upon Discord Player, which is honestly crazy.

Discord Player have been getting some unexpected issues, mainly due to us trying to re-use old code on every new updates. To solve this, what I would like to do with this library for version 6 is:
* Split the discord-player into multiple packages
  currently available:
  * **[@discord-player/core](https://github.com/discord-player/core)**
  * **[@discord-player/extractor](https://github.com/discord-player/extractor)**
  * **[@discord-player/downloader](https://github.com/discord-player/downloader)**
* Better ways to allow users to take control over how things are done
* Learn from past mistakes
* Focus on quality
* Performance improvements, etc.

My ideas might not be enough and I want this community to share their views/ideas for the future updates of this library.
Please leave your ideas below on what you would like to see in the next version of discord-player.

Also the related projects of **[discord-player](https://github.com/Androz2091/discord-player)** will be available on https://github.com/discord-player organization.

**Comment by lailinctrl** [comment]
we need a support server i guess

**Reply by pauldb09** [reply]
There's one: https://discord.gg/gDFaGAGW

**Comment by pauldb09** [comment]
@DevAndromeda Discord-player is really cool and I used it for few months with my bot but it's really not adapted for really big bots, especially with the memory leak when too much users are calling `queue.connect()` in streamDispatcher.
You should also improve the seek function and filter

**Reply by twlite** [reply]
Seek does seem to work somewhat, but since the stream is being re-created using current approach, it is slow (this applies to all other filters). This could be improved

**Comment by twlite** [comment]
**Extractors API**
- should allow interaction with the core module
- should be able to control info/data fetching
- should be able to access/add/modify props on the core module

#### Basic Example

```ts
import { registerExtractor, QueryTypes } from "discord-player"

class MyExtractor {
    constructor() {
        QueryTypes.register("MY_QUERY_TYPE", "someValue");
    }

    validate(src) {
        /* some logic */
    }

    handleQuery(query, metadata) {
        /* return result */
    }
}

registerExtractor("id", MyExtractor);
```

**Plugins API**
- everything from extractors api
- should also be able to control streaming behavior
- should have access to streaming API (example volume controls, etc)

#### Example

```ts
import { registerPlugin, Player } from "discord-player"

class MyExtractor {
    constructor(readonly player: Player) {}

    handle(queue, metadata) {
        /* return result */
    }
}

registerPlugin("id", MyPlugin);
```
